pypiserver/backend.py

```diff
@@ -0,0 +1,224 @@
+import hashlib
+import itertools
+import os
+import typing as t
+from pathlib import Path, PurePath
+
+from . import Configuration
+from .cache import ENABLE_CACHING, CacheManager
+from .pkg_helpers import (
+    normalize_pkgname,
+    parse_version,
+    is_allowed_path,
+    guess_pkgname_and_version,
+)
+
+PathLike = t.Union[str, bytes, Path, PurePath]
+
+
+class PkgFile:
+    __slots__ = [
+        "pkgname",  # The projects/package name with possible capitalization
+        "version",  # The package version as a string
+        "fn",  # The full file path
+        "root",  # An optional root directory of the file
+        "relfn",  # The file path relative to the root
+        "replaces",  # The previous version of the package (used by manage.py)
+        "pkgname_norm",  # The PEP503 normalized project name
+        "digest",  # The file digest in the form of <algo>=<hash>
+        "relfn_unix",  # The relative file path in unix notation
+        "parsed_version",  # The package version as a tuple of parts
+        "digester",  # a function that calculates the digest for the package
+    ]
+
+    def __init__(
+        self, pkgname, version, fn=None, root=None, relfn=None, replaces=None
+    ):
+        self.pkgname = pkgname
+        self.pkgname_norm = normalize_pkgname(pkgname)
+        self.version = version
+        self.parsed_version = parse_version(version)
+        self.fn = fn
+        self.root = root
+        self.relfn = relfn
+        self.relfn_unix = None if relfn is None else relfn.replace("\\", "/")
+        self.replaces = replaces
+        self.digest = None
+
+    def __repr__(self):
+        return "{}({})".format(
+            self.__class__.__name__,
+            ", ".join(
+                [
+                    f"{k}={getattr(self, k, 'AttributeError')!r}"
+                    for k in sorted(self.__slots__)
+                ]
+            ),
+        )
+
+    @property
+    def fname_and_hash(self):
+        if self.digest is None:
+            self.digest = self.digester(self)
+        hashpart = f"#{self.digest}" if self.digest else ""
+        return self.relfn_unix + hashpart
+
+
+class Backend:
+    def __init__(self, config: Configuration):
+        self.hash_algo = config.hash_algo
+
+    def get_all_packages(self) -> t.Iterable[PkgFile]:
+        """Implement this method to return an Iterable of all packages (as
+        PkgFile objects) that are available in the Backend.
+        """
+        raise NotImplementedError
+
+    def add_package(self, filename: str, fh: t.BinaryIO):
```

Comment:
bikeshed0: I think `stream` or `bytestream` may be a more generic name for this parameter and will make more sense for implementations not using the FS

we may also explicitly want to note the return type being `None` here (and for other methods with no return value)
```suggestion
    def add_package(self, filename: str, fh: t.BinaryIO) -> None:
```